lib/mongoid/matchable.rb

```diff
@@ -151,22 +151,37 @@ def matcher(document, key, value)
       # @since 2.2.1
       # @api private
       def extract_attribute(document, key)
-        if (key_string = key.to_s) =~ /.+\..+/
+        if (key_string = key.to_s).include?('.')
+          # The key specified descends into a hash field or an
+          # embedded document.
+          # Traverse the key as a path from the top-level document downward.
           key_string.split('.').inject(document.send(:as_attributes)) do |_attribs, _key|
             if _attribs.is_a?(::Array)
-              if _key =~ /\A\d+\z/ && _attribs.none? {|doc| doc.is_a?(Hash)}
+              # The attributes at this level are an array.
+              no_hash_key_matches = _attribs.none? {|doc| doc.is_a?(Hash) && doc.has_key?(_key)}
+
+              if _key =~ /\A\d+\z/ && no_hash_key_matches
                 _attribs.try(:[], _key.to_i)
+              elsif no_hash_key_matches
+                nil
```

Comment:
This line returns nil for arrays where previously the return value was an array, this needs an explanation and assessment on all other operators and traversal scenarios.

Reply:
Assessment of impact on other operators/scenarios.